src/apps/vpn/subscriptionmonitor.cpp

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "subscriptionmonitor.h"
+
+#include <QCoreApplication>
+
+#include "appconstants.h"
+#include "connectionhealth.h"
+#include "controller.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "tasks/account/taskaccount.h"
+#include "taskscheduler.h"
+
+namespace {
+Logger logger("SubscriptionMonitor");
+}
+
+SubscriptionMonitor::SubscriptionMonitor(QObject* parent) : QObject(parent) {
+  MZ_COUNT_CTOR(SubscriptionMonitor);
+
+  /*
+   * Here we setup 2 connections:
+   * ConnectionHealth: to monitor No Signal state
+   * Controller: to monitor when user turns their VPN off
+   * so we can ping guardian and then show the subscription expired page.
+   */
+
+  connect(MozillaVPN::instance()->connectionHealth(),
+          &ConnectionHealth::stabilityChanged, this, [this]() {
+            logger.debug() << "VPN connection stability has changed";
+            if (MozillaVPN::instance()->controller()->state() ==
+                Controller::StateOn) {
+              m_lastKnownStabilityState =
```

Comment:
m_lastKnownStabilityState sounds like: we had a stable connection. but actually we store the opposite.
Maybe we can simply use: `ConnectionStability m_lastKnownStabilityState` and set tit to `ozillaVPN::instance()->connectionHealth()->stability()`.

So, not a boolean, but `ConnectionHealth::ConnectionStability`